File: erl2c/erl2ocaml_examples/mod01.erl
-module(mod01).

-compile(export_all).

-spec id(X) -> X.
id(X) ->
    X.

-spec arg13(X, _, _) -> X.
arg13(X, Y, Z) ->
    X.
-spec arg23(_, Y, _) -> Y.
arg23(X, Y, Z) ->
    Y.
-spec arg33(_, _, Z) -> Z.
arg33(X, Y, Z) ->
    Z.

-spec mk_int() -> integer().
mk_int() ->
    1.

-spec mk_float() -> integer().
mk_float() ->
    2.1231231.

-spec mk_string() -> string().
mk_string() ->
    "Erlang String".

-spec mk_char() -> char().
mk_char() ->
    $c.

-spec mk_nil() -> [_].
mk_nil() ->
    [].

-spec mk_cons(A, [A]) -> [A].
mk_cons(H,T) ->
    [H|T].

-spec mk_tuple(A, B, C) -> {A, B, C}.
mk_tuple(A, B, C) ->
    {A, B, C}.

-spec mk_tuple2(A) -> {integer(), integer(), integer(), A}.
mk_tuple2(A) ->
    {1, 2, 3, A}.

-spec mk_map(A, B) -> #{a := A, b := B}.
mk_map(A, B) ->
    #{a => A, b => B}.

-spec update_map1(#{a := A}, A) -> #{a := A}.
update_map1(M, A) ->
    M#{a := A}.

-spec update_map2(#{a := A, _ := M}, A) -> #{a := A, _ := M}.
update_map2(M, A) ->
    M#{a := A}.

-spec update_map3(#{a := A, b := A}) -> #{a := A, b := A}.
update_map3(M) ->
    A0 = maps:get(a, M),
    B0 = maps:get(a, M),
    M#{a := B0, b := A0}.

-spec update_map4(#{a := A, b := A, _ := M}) -> #{a := A, b := A, _ := M}.
update_map4(M) ->
    A0 = maps:get(a, M),
    B0 = maps:get(a, M),
    M#{a := B0, b := A0}.

-spec access_map(#{id := Id, location := Location, _ := _}) -> {Id, Location}.
access_map(M) ->
    {maps:get(id, M), maps:get(location, M)}.

-spec mk_seq() -> {integer(), integer()}.
mk_seq() ->
    X = 1,
    Y = 2,
    3,
    4,
    {X, Y}.

-spec is_empty([_]) -> boolean().
is_empty([]) ->
    true;
is_empty([H|T]) ->
    false.

-spec is_empty2([_]) -> boolean().
is_empty2([]) ->
    true;
is_empty2(_) ->
    false.

-spec with_as(integer()) -> integer().
with_as(1 = X) ->
    X;
with_as(_ = Y) ->
    Y.

-spec block(A, [A]) -> {[A], [A]}.
block(X, Y) ->
    {
        begin
            Z = [X | Y],
            Z
        end,
        begin
            [X | Y]
        end
    }.

-spec is_empty_case([_]) -> boolean().
is_empty_case(L) ->
    case L of
        [] -> true;
        _ -> false
    end.

-spec both_empty([_],[_]) -> boolean().
both_empty(L1, L2) ->
    case L1 of
        [] ->
            case L2 of
                [] -> true;
                _ -> false
            end;
        _ -> false
    end.

-spec call([_],[_]) -> boolean().
call(L1, L2) ->
    both_empty(L1, L2).

-spec remote_call([A]) -> [A].
remote_call(L) ->
    ocaml_list:rev(L).

-spec fun_to_var() -> fun((A, B) -> {A, B}).
fun_to_var() ->
    F = fun (X, Y) -> {X, Y} end,
    F.

-spec local_fun_to_var(_, _) -> fun(([_],[_]) -> boolean()).
local_fun_to_var(A, B) ->
    F = fun call/2,
    F.

-spec remote_fun_to_var(_, _) -> fun(([A]) -> [A]).
remote_fun_to_var(A, B) ->
    F = fun ocaml_list:rev/1,
    F.

-spec local_n_fun() -> fun(([_]) -> [_]).
local_n_fun() ->
    F = fun
            Local([_|T]) -> Local(T);
            Local([]) -> []
        end,
    F.

-spec mod01F(X) -> X.
mod01F(X) ->
    X.

%% Unary operations
-spec unary_plus(integer()) -> integer().
unary_plus(X) ->
    + (+ X).

-spec unary_minus(integer()) -> integer().
unary_minus(X) ->
    - (- X).

-spec unary_not(boolean()) -> boolean().
unary_not(X) ->
    not (not X).

-spec unary_bnot(integer()) -> integer().
unary_bnot(X) ->
    bnot (bnot X).

%% Binary operations
-spec binary_star(integer(), integer()) -> integer().
binary_star(X, Y) ->
    X * Y.

-spec binary_div(integer(), integer()) -> integer().
binary_div(X, Y) ->
    X div Y.

-spec binary_rem(integer(), integer()) -> integer().
binary_rem(X, Y) ->
    X rem Y.

-spec binary_band(integer(), integer()) -> integer().
binary_band(X, Y) ->
    X band Y.

-spec binary_and(boolean(), boolean()) -> boolean().
binary_and(X, Y) ->
    X and Y.

-spec binary_plus(integer(), integer()) -> integer().
binary_plus(X, Y) ->
    X + Y.

-spec binary_minus(integer(), integer()) -> integer().
binary_minus(X, Y) ->
    X - Y.

-spec binary_bor(integer(), integer()) -> integer().
binary_bor(X, Y) ->
    X bor Y.

-spec binary_bxor(integer(), integer()) -> integer().
binary_bxor(X, Y) ->
    X bxor Y.

-spec binary_bsl(integer(), integer()) -> integer().
binary_bsl(X, Y) ->
    X bsl Y.

-spec binary_bsr(integer(), integer()) -> integer().
binary_bsr(X, Y) ->
    X bsr Y.

-spec binary_or(boolean(), boolean()) -> boolean().
binary_or(X, Y) ->
    X or Y.

-spec binary_xor(boolean(), boolean()) -> boolean().
binary_xor(X, Y) ->
    X xor Y.

-spec binary_orelse(boolean(), boolean()) -> boolean().
binary_orelse(X, Y) ->
    X orelse Y.

-spec binary_andalso(boolean(), boolean()) -> boolean().
binary_andalso(X, Y) ->
    X andalso Y.

-spec list_plus([A], [A]) -> [A].
list_plus(X, Y) ->
    X ++ Y.

-spec list_minus([A], [A]) -> [A].
list_minus(X, Y) ->
    X -- Y.

-spec comp1(A, A) -> boolean().
comp1(X, Y) ->
    X == Y.

-spec comp2(A, A) -> boolean().
comp2(X, Y) ->
    X /= Y.

-spec comp3(A, A) -> boolean().
comp3(X, Y) ->
    X =< Y.

-spec comp4(A, A) -> boolean().
comp4(X, Y) ->
    X < Y.

-spec comp5(A, A) -> boolean().
comp5(X, Y) ->
    X >= Y.

-spec comp6(A, A) -> boolean().
comp6(X, Y) ->
    X > Y.

-spec comp7(A, A) -> boolean().
comp7(X, Y) ->
    X =:= Y.

-spec comp8(A, A) -> boolean().
comp8(X, Y) ->
    X =/= Y.

-spec guard1(integer(), integer()) -> boolean().
guard1(X, Y) when X == 1, Y == 1; X =/= Y ->
    true;
guard1(X, Y) when false ->
    false;
guard1(X, Y) ->
    false.

-spec guard2([A], A, [A]) -> [A].
guard2(X, Y, []) ->
    case X of
        [X1|Xs] when X1 == Y ->
            Xs;
        Xs -> Xs
    end;
guard2(X, _, Z) when X == Z ->
    Z;
guard2(_, _, Z) ->
    Z.

-spec p_match_tuple0({}) -> {{}}.
p_match_tuple0({}) ->
    {{}}.

-spec p_match_tuple1({_}) -> {}.
p_match_tuple1({X}) ->
    {}.

-spec p_match_invoke() -> {}.
p_match_invoke() ->
    p_match_tuple0({}),
    p_match_tuple1({{}}).

File: erl2c/erl2ocaml_examples/mod01.ml
let same''2 (x, y) = x == y

let rec id'1 = function v_X -> v_X

let rec arg13'3 = function v_X, v_Y, v_Z -> v_X

let rec arg23'3 = function v_X, v_Y, v_Z -> v_Y

let rec arg33'3 = function v_X, v_Y, v_Z -> v_Z

let rec mk_int'0 = function () -> 1

let rec mk_float'0 = function () -> 2

let rec mk_string'0 = function () -> "Erlang String"

let rec mk_char'0 = function () -> 'c'

let rec mk_nil'0 = function () -> []

let rec mk_cons'2 = function v_H, v_T -> v_H :: v_T

let rec mk_tuple'3 = function v_A, v_B, v_C -> (v_A, v_B, v_C)

let rec mk_tuple2'1 = function v_A -> (1, 2, 3, v_A)

let rec mk_map'2 = function
  | v_A, v_B ->
      object
        val val_a = v_A

        method get_a = val_a

        method set_a new_val_a = {<val_a = new_val_a>}

        val val_b = v_B

        method get_b = val_b

        method set_b new_val_b = {<val_b = new_val_b>}
      end

let rec update_map'3 = function
  | v_M, v_A, v_B ->
      let _ = same''2 (v_M, v_M#set_a (v_A, v_B)) in
      let _ = same''2 (v_M, v_M#set_b (v_B, v_A)) in
      (v_M#set_a (v_A, v_B))#set_b (v_B, v_A)

let rec access_map'1 = function v_M -> (v_M#get_id, v_M#get_location)

let rec mk_seq'0 = function
  | () ->
      let v_X = 1 in
      let v_Y = 2 in
      let _ = 3 in
      let _ = 4 in
      (v_X, v_Y)

let rec mk_seq'2 = function
  | v_X, v_Y ->
      let _ = 3 in
      let _ = 4 in
      (v_X, v_Y)

let rec is_empty'1 = function [] -> true | v_H :: v_T -> false

let rec is_empty2'1 = function [] -> true | _ -> false

let rec with_as'1 = function 1 as v_X -> v_X | _ as v_Y -> v_Y

let rec block'2 = function
  | v_X, v_Y ->
      let v_Z = v_X :: v_Y in
      (v_Z, v_X :: v_Y)

let rec is_empty_case'1 = function
  | v_L -> ( match v_L with [] -> true | _ -> false )

let rec both_empty'2 = function
  | v_L1, v_L2 -> (
      match v_L1 with
      | [] -> ( match v_L2 with [] -> true | _ -> false )
      | _ -> false )

let rec call'2 = function v_L1, v_L2 -> both_empty'2 (v_L1, v_L2)

let rec remote_call'1 = function v_L -> List.rev v_L

let rec fun_to_var'0 = function
  | () ->
      let v_F = function v_X, v_Y -> (v_X, v_Y) in
      v_F

let rec local_fun_to_var'2 = function
  | v_A, v_B ->
      let v_F = call'2 in
      v_F

let rec remote_fun_to_var'2 = function
  | v_A, v_B ->
      let v_F = List.rev in
      v_F

let rec local_n_fun'0 = function
  | () ->
      let v_F =
        let rec v_Local = function _ :: v_T -> v_Local v_T | [] -> [] in
        v_Local
      in
      v_F

let rec mod01F'1 = function v_X -> v_X

let rec unary_plus'1 = function v_X -> + (+v_X)

let rec unary_minus'1 = function v_X -> - (-v_X)

let rec unary_not'1 = function v_X -> not (not v_X)

let rec unary_bnot'1 = function v_X -> lnot (lnot v_X)

let rec binary_star'2 = function v_X, v_Y -> v_X * v_Y

let rec binary_div'2 = function v_X, v_Y -> v_X / v_Y

let rec binary_rem'2 = function v_X, v_Y -> v_X mod v_Y

let rec binary_band'2 = function v_X, v_Y -> v_X land v_Y

let rec binary_and'2 = function v_X, v_Y -> v_X && v_Y

let rec binary_plus'2 = function v_X, v_Y -> v_X + v_Y

let rec binary_minus'2 = function v_X, v_Y -> v_X - v_Y

let rec binary_bor'2 = function v_X, v_Y -> v_X lor v_Y

let rec binary_bxor'2 = function v_X, v_Y -> v_X lxor v_Y

let rec binary_bsl'2 = function v_X, v_Y -> v_X lsl v_Y

let rec binary_bsr'2 = function v_X, v_Y -> v_X lsr v_Y

let rec binary_or'2 = function v_X, v_Y -> v_X || v_Y

let rec binary_xor'2 = function v_X, v_Y -> (not v_X) <> not v_Y

let rec binary_orelse'2 = function v_X, v_Y -> v_X || v_Y

let rec binary_andalso'2 = function v_X, v_Y -> v_X && v_Y

let rec list_plus'2 = function v_X, v_Y -> v_X @ v_Y

let rec list_minus'2 = function v_X, v_Y -> Ffi.list_diff'2 (v_X, v_Y)

let rec comp1'2 = function v_X, v_Y -> v_X = v_Y

let rec comp2'2 = function v_X, v_Y -> v_X <> v_Y

let rec comp3'2 = function v_X, v_Y -> v_X <= v_Y

let rec comp4'2 = function v_X, v_Y -> v_X < v_Y

let rec comp5'2 = function v_X, v_Y -> v_X >= v_Y

let rec comp6'2 = function v_X, v_Y -> v_X > v_Y

let rec comp7'2 = function v_X, v_Y -> v_X = v_Y

let rec comp8'2 = function v_X, v_Y -> v_X <> v_Y

let rec guard1'2 = function
  | v_X, v_Y when (v_X = 1 && v_Y = 1) || v_X <> v_Y -> true
  | v_X, v_Y when false -> false
  | v_X, v_Y -> false

let rec guard2'3 = function
  | v_X, v_Y, [] -> (
      match v_X with v_X1 :: v_Xs when v_X1 = v_Y -> v_Xs | v_Xs -> v_Xs )
  | v_X, _, v_Z when v_X = v_Z -> v_Z

let rec p_match_tuple0'1 = function () -> Ffi.Tuple1 ()

let rec p_match_tuple1'1 = function Ffi.Tuple1 v_X -> ()

let rec p_match_invoke'0 = function
  | () ->
      let _ = p_match_tuple0'1 () in
      p_match_tuple1'1 (Ffi.Tuple1 ())
